include/Signal.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+#include <memory>
+#include <vector>
+#include <functional>
+
+namespace Framework
+{
+
+	template<typename> class CSignal;
+	template<typename T, typename... Args>
+	class CSignal<T (Args...)>
+	{
+
+	public:
+		class CConnection;
+
+		typedef std::shared_ptr<CConnection> CConnectionPtr;
+		typedef std::weak_ptr<CConnection> CConnectionWeakPtr;
+		typedef std::vector<CConnectionPtr> CConnectionListPtr;
+
+		typedef CSignal<T (Args...)> CSignalProto;
+		typedef std::function<void()> CConectionFunction;
+		typedef std::function<T (Args...)> CSlotFunction;
+		typedef std::weak_ptr<const CSlotFunction> CSlotFunctionPtr;
+
+		CSignal() = default;
+
+		CConnectionPtr Connect(const CSlotFunction& func, bool oneShot = false)
+		{
+			if(func)
```

Comment:
I think we should just use an `assert` here. Providing a non-empty function pointer is a precondition, it shouldn't be something that is provided by mistake and we don't want to provide error checking for this either.